design/open/high-level-python-wrapper.md

```diff
@@ -0,0 +1,314 @@
+
+Additionally, we define the following exception hierarchy:
+
+```python
+
+# lakefs.exceptions
+class LakeFSException(Exception):
+    status_code: int
+    message: str
+
+class NotFoundException(LakeFSException): ...
+class NotAuthorizedException(LakeFSException): ...
+class ServerException(LakeFSException): ...
+class UnsupportedOperationException(LakeFSException): ...
+```
+
+Other, more specific exceptions may subclass these, but all errors returned by the lakeFS server should sub-class one of these to make error handling easier for developers.
+
+The only exception should be errors returned by functions that, for compatibility should return specific error types. (this might be true for e.g. `open()` and errors returned by the `TextIO` or `BinaryIO` implementations).
+
+## Higher Level Utilities
+
+### I/O - reading/writing objects
+
+Provide a pythonic `open()` method that returns a "file-like object", passable to other frameworks or libraries
+that can deal with one.
+
+```python
+import lakefs
+
+repo = lakefs.repository('example')
+branch = repo.branch('main')
+
+data: bytes = b'hello world\n'
+with branch.open('foo/bar.txt', 'w') as f:
+    # Will check the underlying client for pre-signed URL support
+    #  if supported, will do the whole get_physical_address, http upload, link address dance
+    #  Otherwise, direct upload.
+    #  *In the future*, we can accept a stream/file-like object, sniff for its size/content type
+    #  opt for multi-part, etc.
+    f.write(data)
+    f.write(another_thing)  # fail this, we can only write once.
+
+with repo.commit('abc123').open('foo/bar.txt', 'r') as f:
+    # same thing - check client for pre-signed support
+    # fall back to direct read if not supported
+    f.read()
+    # or
+    f.read(1024)  # range request
+
+# 
+# 
+```
+
+`open()` will also accept an explicit `pre_signed: Optional[bool] = None` argument.
+if set, don't try and probe the client for this capability
+
+### Import Manager
+
+Provide a utility to run 
+
+```python
+import lakefs
+
+main = lakefs.repository('example').branch('main')
+task = main.import_from(
+    's3://bucket/path', 
+    destination='some/path/',
+    commit_message='imported stuff!'
+)
+task.start()  # will not block, run the import API
+task.wait()  # Block, polling in the background
+
+# or just run(), same as start() & wait()
+main.import_from(object_store_uri='s3://bucket/path', destination='datasets/', commit_message='sync datasets').run()
+```
+
+### Transaction Manager
+
+```python
+import lakefs
+
+dev = lakefs.repository('example').branch('dev')
+
+# Will create an ephemeral branch from `dev` (e.g. `tx-dev-343829f89`)
+# uploads and downloads will apply to that ephemeral branch
+# on success, commit with provided message, merge and delete ephemeral branch
+# on exception or failure, leave branch as is and report it in a wrapping exception
+# for easy troubleshooting
```

Comment:
This is beautiful, very much what we designed lakeFS to be!

Worried about pushing this feature on the client-side before we build the server-side to make it usable in real-world scenarios.

We currently have poor support for multiple concurrent branch merges (even when they do not conflict), and additionally they may work extremely slowly.  I understand the desire to build demand ahead of the feature, but it is vital to understand that there will be no quick fixes for these.

I believe that we need to push people to use lakeFS, which means merges.  (This is emphatically _not_ the "dev" workflow that we've been pursuing.)  But I don't want them to be disappointed.  So while we can defer _functional_ requirements, I believe we should prioritize _nonfunctional_ requirements notably performance.

# Performance changes to prioritize

Any slowness in merge compounds exponentially: if merges take 2x as long, the probability of 2 concurrent merges grows approximately 2x, which means that the losing merge needs to be retried, so it actually takes 4x as long, ...  If merges occur at a roughly constant rate and are retried, then there is a critical rate at which a naïve implementation will occasionally block and never recover.

* #2795 to make merges work faster
* [Reuse metaranges to retry contended merges](https://github.com/treeverse/lakeFS/blob/master/design/open/reuse-contented-merge-metaranges/reuse-contented-merge-metaranges.md) can specifically speed up the retry operation during merges.
* #5613 includes making find-merge-base faster.  It used to be slow, and I believe that it will still turn out to be too slow.

### Merge API changes that may be deferred

* #3639 Per-Object Merge Strategy: This feature might be extended to allow Pythonic code to fix the merge.  But it's actually quite hard to do without undoing all the changes performed by the transaction.  And if retries are slow, any races will cause unbounded and large latency.
* #3274, #3316 to help with fast-forward merges: This feature may change semantics of `dev.transaction`.